proto/kv-connect.md

```diff
@@ -0,0 +1,448 @@
+database metadata expires. The client MUST NOT use the returned database
+metadata after this date and time. The client SHOULD cache the returned database
+metadata and use it until it expires. The client SHOULD perform a metadata
+exchange with the server just before the metadata expires to ensure that the
+client is using the latest database metadata.
+
+Example metadata exchange:
+
+```http
+POST / HTTP/1.1
+Authorization: Bearer a1b2c3d4e5f6g7h8i9
+Content-Type: application/json
+Content-Length: 33
+
+{
+  "supportedVersions": [1, 2]
+}
+```
+
+```http
+HTTP/1.1 200 OK
+Content-Type: application/json
+Content-Length: 315
+
+{
+  "version": 2,
+  "uuid": "a1b2c3d4-e5f6-7g8h-9i1j-2k3l4m5n6o7p",
+  "endpoints": [
+    {
+      "url": "/v1",
+      "consistency": "strong"
+    },
+    {
+      "url": "https://mirror.example.com/v1",
+      "consistency": "eventual"
+    }
+  ],
+  "token": "123abc456def789ghi",
+  "expiresAt": "2023-10-01T00:00:00Z"
+}
+```
+
+### Data Path Protocol
+
+After performing the metadata exchange, the client can perform CRUD operations
+on the KV database using the _Data Path Protocol_. The _Data Path Protocol_ is a
+Protobuf over HTTP protocol.
+
+All requests on this protocol MUST include an `Authorization` header with the
+value `Bearer <token>`, where `<token>` is the token returned by the server
+during the metadata exchange.
+
+The client MUST include on all requests a `User-Agent` header with a value
+identifying the client library and version that is being used.
+
+The client MUST include on all requests a `Content-Type` header with the value
+`application/x-protobuf`.
+
+If the protocol version used is 1, the client MUST include on all requests a
+`x-transaction-domain-id` header with a value of the UUID of the database as
+returned by the server during the metadata exchange.
+
+If the protocol version used is 2 or higher, the client MUST include on all
+requests a `x-denokv-database-id` header with a value of the UUID of the
+database as returned by the server during the metadata exchange.
+
+If the protocol version used is 2 or higher, the client MUST include on all
+requests a `x-denokv-version` header with a value of the protocol version that
+the client is using.
+
+The server MUST NOT respond with 3xx class redirect responses to data path
+requests.
+
+There are two types of requests that the client may send to the server:
+
+- A _Snapshot Read Request_ which is used to read some values from the database.
+- An _Atomic Write Request_ which is used to write some values to the database.
+
+#### Snapshot Read Request
+
+A _Snapshot Read Request_ is used to read some values from the database.
+
+<!-- A _Snapshot Read Request_ can either be "single shot" or "watching". A "single
+shot" _Snapshot Read Request_ is used to read a given set of values from the
+database once. A "watching" _Snapshot Read Request_ is used to read a given set
+of values from the database once and then receive a stream of  -->
```

Comment:
There are a few commented out paragraphs. Did you mean to leave them in?

Reply:
Watch stuff is not quite ready yet - Heyang and me are still debating on the final API, so I left this out for the moment.